### PR TITLE
More precise newsflash coordinates (using Waze data)

### PR DESCRIPTION
**Pull request by sahare92**
Added a simple logic to find the related waze accident alert using information from the newsflash
(Currently it selects a waze accident that occurred around the area of the newsflash accident , up to `WAZE_ALERT_NEWSFLASH_DELTA_IN_HOURS` hours before the newsflash's time.

It currently only populates the `newsflash.waze_alert` property, without modifying the lat/lon attributes according to the waze accident (this should be updated after we test it manually, and make sure it indeed improves the precision)

### CODE REVIEW

File: anyway/parsers/location_extraction.py [hunk 1]
```diff
@@ -297,15 +301,60 @@ def extract_location_text(text):
...
+
+    # determine what distance (in kilometers) to look for waze accidents in, according to the newsflash's resolution
+    if newsflash.resolution in short_distance_resolutions:
+        distance = 0.3
+    elif newsflash.resolution in long_distance_resolutions:
+        distance = 5
+    else:
+
+        # unknown resolution - skip this optimization
+        return None
```
Comment (elazarg) [inline anyway/parsers/location_extraction.py:314]: There's a stray newline here.

Consider making `distance_resolution` a dictionary, mapping names to distances, on which you can iterate. Just a suggestion.

File: anyway/parsers/utils.py [hunk 2]
```diff
@@ -14,3 +17,27 @@ def batch_iterator(iterable, batch_size):
...
+    radius = 6371
+    # Radius of the parallel at given latitude
+    parallel_radius = radius * math.cos(latitude)
+
+    lat_min = latitude - distance_in_km / radius
+    lat_max = latitude + distance_in_km / radius
+    lon_min = longitude - distance_in_km / parallel_radius
+    lon_max = longitude + distance_in_km / parallel_radius
+    rad2deg = math.degrees
+
```
Comment (elazarg) [inline anyway/parsers/utils.py:35]: Maybe move this newline one line upwards :)

File: anyway/parsers/utils.py [hunk 3]
```diff
@@ -14,3 +17,27 @@ def batch_iterator(iterable, batch_size):
...
+    baseX = rad2deg(lon_min)
+    baseY = rad2deg(lat_min)
+    distanceX = rad2deg(lon_max)
+    distanceY = rad2deg(lat_max)
+
+    return "POLYGON(({0} {1},{0} {3},{2} {3},{2} {1},{0} {1}))".format(
+        baseX, baseY, distanceX, distanceY
+    )
```
Comment (elazarg) [inline anyway/parsers/utils.py:43]: Maybe use an f-string?
```python
return f"POLYGON(({x} {y},{x} {dy},{dx} {dy},{dx} {y},{x} {y}))"
```

File: anyway/parsers/waze/waze_data_parser.py [hunk 4]
```diff
@@ -73,9 +73,11 @@ def parse_waze_alerts_data(waze_alerts):
     )
     waze_df["road_type"] = int(waze_df["road_type"].fillna(-1)[0])
     waze_df["number_thumbs_up"] = int(waze_df.get("number_thumbs_up").fillna(0)[0])
-    waze_df["report_by_municipality_user"] = _convert_to_bool(waze_df.get("report_by_municipality_user", False))
+    waze_df["report_by_municipality_user"] = _convert_to_bool(
+        waze_df.get("report_by_municipality_user", False)
+    )
```
Comment (elazarg) [inline anyway/parsers/waze/waze_data_parser.py]: I wonder, why does black change these lines? How could they have been fine so far?

File: anyway/parsers/waze/waze_db_functions.py [hunk 5]
```diff
@@ -25,7 +25,9 @@ def _upsert_waze_objects_by_uuid(model, waze_objects):
     with db.session.no_autoflush:
         for waze_object in waze_objects:
             db.session.flush()
-            existing_objects = db.session.query(model).filter(model.uuid == str(waze_object["uuid"]))
+            existing_objects = db.session.query(model).filter(
+                model.uuid == str(waze_object["uuid"])
+            )
```
Comment (elazarg) [inline anyway/parsers/waze/waze_db_functions.py]: Again, do you happen to know why does black touch this file?

File: tests/test_news_flash.py [hunk 6]
```diff
@@ -268,3 +320,40 @@ def test_classification_statistics_ynet():
     assert precision > BEST_PRECISION_YNET
     assert recall > BEST_RECALL_YNET
     assert f1 > BEST_F1_YNET
+
+
```
Comment (elazarg) [inline tests/test_news_flash.py:324]: There's a better and more idiomatic way of coding this pattern. Use [contextlib.contextmanager](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager). Roughly:

```python
@contextmanager
def insert_waze_alert_to_db():
    try:
        ... # create accident alert
        yield waze_alert
    finally:
        ... # delete waze alert
```

Client code will use a `with` statement.

Actually, and probably even better, would be using pytest's [fixtures](https://docs.pytest.org/en/stable/fixture.html#fixtures-as-function-arguments) for that. But no other test use them, so I'd put that in a TODO.

**Review by elazarg — COMMENTED** [review]
Nice work! I have added some comments, but they are mostly minor suggestions.

I am not entirely sure the best way of adding SQL rows for tests is using this "add, commit; use; delete, commit" pattern; I'm not an expert, but I think it could be done using something like "add; use; rollback" pattern. Maybe @ziv17 would like to comment on this.

File: tests/test_news_flash.py (hunk 6 above)
Comment (ziv17) [reply tests/test_news_flash.py:324]: Looking at this, I have an unrelated Python question: Assuming we are not using threads, why use `yield`?

**Comment by ziv17** [comment]
> ziv
You can mock the db to return what you need for the test. Using mock you get much faster tests, and it is easier to test for boundary conditions. The tests can run the tests directly from the IDE.

File: tests/test_news_flash.py (hunk 6 above)
Comment (elazarg) [reply tests/test_news_flash.py:324]: Yield in python has nothing to do with threading. It's about separating different control flow logics. It is a concise way of writing iterator objects (generators) using function-like syntax. In this example it simply splits the function to two phases, "enter" and "exit".

**Comment by sahare92** [comment]
@ziv17 I think there's a benefit to run the query on the db (instead of mocking), because this way we can test that the generated query is working, and actually returns the expected WazeAlert from the db.
but I can mock the db response if you think it'll be better. WDYT?

**Comment by elazarg** [comment]
I agree with Ziv. Mocking is generally better, and it's how other tests are written. If you want to test querying the DB, that would be a different test.

File: tests/test_news_flash.py [hunk 7]
```diff
@@ -268,3 +317,49 @@ def test_classification_statistics_ynet():
...
+@contextmanager
+def _managed_waze_accident_alert():
+    waze_alert = _create_waze_accident_alert()
+    try:
+        yield waze_alert
+    finally:
+        _delete_waze_alert(waze_alert.id)
+
+
+def _create_waze_accident_alert():
```
Comment (elazarg) [inline tests/test_news_flash.py:331]: Now that you're using the context manager, there's no need to write these two functions separately. Just inline them.

File: anyway/parsers/location_extraction.py [hunk 8]
```diff
@@ -297,15 +301,56 @@ def extract_location_text(text):
...
+        geo_location["lat"], geo_location["lon"], distance
+    )
+
+    # find waze alerts in that bounding box, from the recent time delta - and return the first as the related waze alert
+    matching_alert = (
+        db.session.query(WazeAlert)
+        .filter(WazeAlert.alert_type == "ACCIDENT")
+        .filter(
+            WazeAlert.created_at.between(
+                newsflash.date - timedelta(hours=WAZE_ALERT_NEWSFLASH_DELTA_IN_HOURS),
```
Comment (elazarg) [inline anyway/parsers/location_extraction.py:323]: Since this constant is only used inside the function, it's should probably be local constant.  This way the occasional reader doesn't have to wonder where else it is used.

Comment (elazarg) [reply anyway/parsers/location_extraction.py:323]: Also, it can be a `timedelta`, making it slightly more type-safe, and making this expression a single line.

**Comment by ziv17** [comment]
> 
> 
> @ziv17 I think there's a benefit to run the query on the db (instead of mocking), because this way we can test that the generated query is working, and actually returns the expected WazeAlert from the db.
> but I can mock the db response if you think it'll be better. WDYT?

We do need both end-to-end tests and tests to cover the logic. Covering the logic with end-to-end will be slow, unstable, and cumbersome. The best practice is small amount of end-to-end and and reach coverage with unit tests (that use mocking)

**Comment by Mano3** [comment]
Hi @sahare92 , great work! 
I'm mostly concerned about the algorithmic aspect.
How were the distances per resolution picked?
How was the time delta picked?
Why taking the first waze alert in the bounding box instead of the closest?
Was this logic checked on a large chunk of existing news flash?
I wish to know some statistics and results of manual review of how this logic fares on a large amount of data before merging.
Thanks.

**Comment by sahare92** [comment]
@Mano3 thanks,

> How were the distances per resolution picked?
How was the time delta picked?
- The distances by resolution were arbitrarily picked  for now, until we'll learn what the best values are.
- The same for time delta, and bounding box
> Why taking the first waze alert in the bounding box instead of the closest?
- We didn't take the closest, because it's not necessarily the right one (for example if the resolution is "עיר" then the geo_location we got from google is just the center of the city, so choosing the closest waze accident doesn't mean it's the right one..but mostly we had hoped that in real data, most times we won't get more than one matching waze accidents (in the time range we will decide to query)
> Also, there is no need to query Waze alert from the news flash datetime minus delta until now. News flash will probably appear after the event is already finished so at most it should query up until news flash datetime + delta.
- Good point, I'll fix that 👍
> Was this logic checked on a large chunk of existing news flash? .......
- No, we didn't test it yet, after talking with @atalyaalon, we thought it'll be a good idea to test this with previous data (hopefully soon, when waze data backfill will be fixed)
Do you have an idea to test this, other then going manually over the data and see if the chosen waze location was accurate according to the newsflash description? if not, then I guess that's what I'll do once waze data backfill is in
